### ISSUE TITLE
Nested differentiation works but the LLVM code looks... really bad

### ISSUE THREAD

**Issue by DilumAluthge** [FluxML/Zygote.jl#253]
TL;DR: In a very simple example, taking the second derivative "works" (in that it will give you the correct output). Howevert, the LLVM code for the second derivative looks awful, especially given that the second derivative is constant - I would have expected Zygote to realize that. Also, running `@code_warntype` shows multiple non-leaf types in the code for the second derivative.

# MWE

Here is the MWE. I ran this on `Zygote#master` and two versions of Julia (`Version 1.1.1 (2019-05-16)` and `Version 1.2.0-rc1.0 (2019-05-30)`).

```julia
using Zygote

f(x) = 5x + 3
f(10), f'(10)
@code_llvm debuginfo=:none f(10)
@code_llvm debuginfo=:none f'(10)

g(x) = 7x^2 + 5x + 3
g(10), g'(10), g''(10)
@code_llvm debuginfo=:none g(10)
@code_llvm debuginfo=:none g'(10)
@code_llvm debuginfo=:none g''(10)
```

# Full output of MWE

```julia
julia> using Zygote

julia> f(x) = 5x + 3
f (generic function with 1 method)

julia> f(10), f'(10)
(53, 5)

julia> @code_llvm debuginfo=:none f(10)

define i64 @julia_f_16083(i64) {
top:
  %1 = mul i64 %0, 5
  %2 = add i64 %1, 3
  ret i64 %2
}

julia> @code_llvm debuginfo=:none f'(10)

define i64 @"julia_#34_16850"(i64) {
top:
  ret i64 5
}

julia> g(x) = 7x^2 + 5x + 3
g (generic function with 1 method)

julia> g(10), g'(10), g''(10)
(753, 145, 14)

julia> @code_llvm debuginfo=:none g(10)

define i64 @julia_g_16881(i64) {
top:
  %1 = mul i64 %0, 7
  %reass.add = add i64 %1, 5
  %reass.mul = mul i64 %reass.add, %0
  %2 = add i64 %reass.mul, 3
  ret i64 %2
}

julia> @code_llvm debuginfo=:none g'(10)

define i64 @"julia_#34_16938"(i64) {
top:
  %1 = mul i64 %0, 14
  %2 = add i64 %1, 5
  ret i64 %2
}

julia> @code_llvm debuginfo=:none g''(10)

define nonnull %jl_value_t addrspace(10)* @"julia_#34_17204"(i64) {
top:
  %1 = alloca %jl_value_t addrspace(10)*, i32 3
  %gcframe = alloca %jl_value_t addrspace(10)*, i32 3
  %2 = bitcast %jl_value_t addrspace(10)** %gcframe to i8*
  call void @llvm.memset.p0i8.i32(i8* %2, i8 0, i32 24, i32 0, i1 false)
  %3 = call %jl_value_t*** inttoptr (i64 4332153408 to %jl_value_t*** ()*)() #2
  %4 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %gcframe, i32 0
  %5 = bitcast %jl_value_t addrspace(10)** %4 to i64*
  store i64 2, i64* %5
  %6 = getelementptr %jl_value_t**, %jl_value_t*** %3, i32 0
  %7 = load %jl_value_t**, %jl_value_t*** %6
  %8 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %gcframe, i32 1
  %9 = bitcast %jl_value_t addrspace(10)** %8 to %jl_value_t***
  store %jl_value_t** %7, %jl_value_t*** %9
  %10 = bitcast %jl_value_t*** %6 to %jl_value_t addrspace(10)***
  store %jl_value_t addrspace(10)** %gcframe, %jl_value_t addrspace(10)*** %10
  %11 = call nonnull %jl_value_t addrspace(10)* @julia_gradient_17205(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 4406991208 to %jl_value_t*) to %jl_value_t addrspace(10)*), i64 %0)
  %12 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %gcframe, i32 2
  store %jl_value_t addrspace(10)* %11, %jl_value_t addrspace(10)** %12
  %13 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %1, i32 0
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 4549153488 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %13
  %14 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %1, i32 1
  store %jl_value_t addrspace(10)* %11, %jl_value_t addrspace(10)** %14
  %15 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %1, i32 2
  store %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 4455980624 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** %15
  %16 = call nonnull %jl_value_t addrspace(10)* @jl_apply_generic(%jl_value_t addrspace(10)** %1, i32 3)
  %17 = getelementptr %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %gcframe, i32 1
  %18 = load %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %17
  %19 = getelementptr %jl_value_t**, %jl_value_t*** %3, i32 0
  %20 = bitcast %jl_value_t*** %19 to %jl_value_t addrspace(10)**
  store %jl_value_t addrspace(10)* %18, %jl_value_t addrspace(10)** %20
  ret %jl_value_t addrspace(10)* %16
}
```

# Discussion

`f'(x) = 5` is a constant function, and Zygote correctly recognizes this. Therefore, the LLVM code for `f'(10)` is very simply: it just returns `5`.

`g''(x) = 14` is also a constant function. Therefore, I expected that Zygote would understand this and generate LLVM code for `g''(10)` that is very simple and just returns `14`. Instead, the LLVM code that Zygote produces for `g''(10)` looks terrible and not at all like what I expected.

# Potential type instability in `g''(10)`?

If you replace `@code_llvm` with `@code_warntype`, you get the following MWE:

```julia
using Zygote

f(x) = 5x + 3
f(10), f'(10)
@code_warntype debuginfo=:none f(10)
@code_warntype debuginfo=:none f'(10)

g(x) = 7x^2 + 5x + 3
g(10), g'(10), g''(10)
@code_warntype debuginfo=:none g(10)
@code_warntype debuginfo=:none g'(10)
@code_warntype debuginfo=:none g''(10)
```

Which when run, produces the following output:

```julia
julia> using Zygote

julia> f(x) = 5x + 3
f (generic function with 1 method)

julia> f(10), f'(10)
(53, 5)

julia> @code_warntype debuginfo=:none f(10)
Variables
  #self#::Core.Compiler.Const(f, false)
  x::Int64

Body::Int64
1 ─ %1 = (5 * x)::Int64
│   %2 = (%1 + 3)::Int64
└──      return %2

julia> @code_warntype debuginfo=:none f'(10)
Variables
  #self#::Core.Compiler.Const(getfield(Zygote, Symbol("##34#35")){typeof(f)}(f), false)
  x::Int64

Body::Int64
1 ─ %1 = Core.getfield(#self#, :f)::Core.Compiler.Const(f, false)
│   %2 = Zygote.gradient(%1, x)::Tuple{Int64}
│   %3 = Base.getindex(%2, 1)::Int64
└──      return %3

julia> g(x) = 7x^2 + 5x + 3
g (generic function with 1 method)

julia> g(10), g'(10), g''(10)
(753, 145, 14)

julia> @code_warntype debuginfo=:none g(10)
Variables
  #self#::Core.Compiler.Const(g, false)
  x::Int64

Body::Int64
1 ─ %1 = Core.apply_type(Base.Val, 2)::Core.Compiler.Const(Val{2}, false)
│   %2 = (%1)()::Core.Compiler.Const(Val{2}(), false)
│   %3 = Base.literal_pow(Main.:^, x, %2)::Int64
│   %4 = (7 * %3)::Int64
│   %5 = (5 * x)::Int64
│   %6 = (%4 + %5 + 3)::Int64
└──      return %6

julia> @code_warntype debuginfo=:none g'(10)
Variables
  #self#::Core.Compiler.Const(getfield(Zygote, Symbol("##34#35")){typeof(g)}(g), false)
  x::Int64

Body::Int64
1 ─ %1 = Core.getfield(#self#, :f)::Core.Compiler.Const(g, false)
│   %2 = Zygote.gradient(%1, x)::Tuple{Int64}
│   %3 = Base.getindex(%2, 1)::Int64
└──      return %3

julia> @code_warntype debuginfo=:none g''(10)
Variables
  #self#::Core.Compiler.Const(getfield(Zygote, Symbol("##34#35")){getfield(Zygote, Symbol("##34#35")){typeof(g)}}(getfield(Zygote, Symbol("##34#35")){typeof(g)}(g)), false)
  x::Int64

Body::Any
1 ─ %1 = Core.getfield(#self#, :f)::Core.Compiler.Const(getfield(Zygote, Symbol("##34#35")){typeof(g)}(g), false)
│   %2 = Zygote.gradient(%1, x)::Union{Nothing, Tuple}
│   %3 = Base.getindex(%2, 1)::Any
└──      return %3
```

`f(10)`, `f'(10)`, `g(10)`, and `g'(10)` are all type-stable. But running `@code_warntype debuginfo=:none g''(10)` shows three non-leaf types:
- `Any`
- `Union{Nothing, Tuple}`
- `Any`

**Comment by MikeInnes** [comment]
> I would have expected Zygote to realize that.

It's important to understand that Zygote doesn't do any kind of simplification or recognise constant derivatives in the way you might expect. It just does a purely mechanical code transformation (which inherently makes code ~2x more complex) which will calculate the correct gradient, possibly while doing other redundant work.

If you're lucky, everything might get inlined and optimised away by the compiler, and you get back `constant 5` or something, but this is currently relatively fragile; there are about 160 functions being optimised away there and only one of them needs to fail the inlining heuristic to prevent global optimisation, and then you see LLVM code for the `gradient` wrapper rather than the actual numerical work. This isn't even necessarily an indicator of poor performance, but what we'd expect from general Julia code.

There are multiple threads of work going on to make this kind of optimisation more reliable, both in terms of inference precision (the `Union{T,Nothing}`s that quickly get promoted to `Any`s) and optimisation across closure boundaries (see the YAKC work). Beyond that, there's not a whole lot to be done on the Zygote side, at least with the tools Julia provides currently.

**Comment by DilumAluthge** [comment]
Thanks Mike! That was a really informative answer.

> If you're lucky, everything might get inlined and optimised away by the compiler, and you get back constant 5 or something, but this is currently relatively fragile; there are about 160 functions being optimised away there and only one of them needs to fail the inlining heuristic to prevent global optimisation, and then you see LLVM code for the gradient wrapper rather than the actual numerical work. This isn't even necessarily an indicator of poor performance, but what we'd expect from general Julia code.

Is there a way for me to explore where in the process the inlining heuristics break down?

For example, I often use the `@code_typed` and `@code_warntype` macros to figure out what the compiler is thinking with regards to type inference and to identify potential sources of type instability. Are there any similar tools that I can use to understanding the inlining decisions being made by the compiler at each level of function call?

**Comment by DilumAluthge** [comment]
For example, I'd be interested in identifying the one or more functions that failed the inlining heuristic in this example.

> There are about 160 functions being optimised away there and only one of them needs to fail the inlining heuristic to prevent global optimisation

**Comment by jpsamaroo** [comment]
https://github.com/JuliaDebug/Cthulhu.jl

**Comment by DilumAluthge** [comment]
Cthulhu is great! I've been experimenting with it a little bit.
